### PR TITLE
[WIP] Enable GET method for functions - proxy fix

### LINKED ISSUE #320: Proposal: Adding mount options to functions.
* Brief summary including motivation/context

The proposed change would allow functions to mount volumes and other directories through the normal docker configuration. This would allow a function to process relatively large amounts of data without having to pass it through http/stdin.

Any design changes

Add docker mount struct to the CreateFunctionRequest struct and passing it along in the  create function handler

Pros + Cons

Pros:

* Functions will be able to have volumes.

Cons:

* Volumes should be available on all nodes running functions (but that is rather expected)

Effort required
Little, it's a two line change.

### PR DESCRIPTION
**Pull request by JockDaRock**
Signed-off-by: Jock Reed jock.da.rock@gmail.com

## Description

Enable passing of the GET method Alex added to enable GET methods.

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))

## How Has This Been Tested?

Tested this in browser and using Postman. Inspected environment variables from flask image as well.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by JockDaRock** [comment]
thoughts on this @alexellis?  😄

File: gateway/handlers/proxy.go [hunk 1]
```diff
@@ -140,7 +141,8 @@ func invokeService(w http.ResponseWriter, r *http.Request, metrics metrics.Metri
 	contentType := r.Header.Get("Content-Type")
 	fmt.Printf("[%s] Forwarding request [%s] to: %s\n", stamp, contentType, url)
 
-	request, err := http.NewRequest("POST", url, bytes.NewReader(requestBody))
+	method := r.Method
+	request, err := http.NewRequest(method, url, bytes.NewReader(requestBody))
```
Comment (alexellis) [inline gateway/handlers/proxy.go]: We can inline r.Method

**Comment by alexellis** [comment]
How much edge-case testing have you done?

* What happens with a function that expects a body but then doesn't get one?
* Is there a difference between Swarm and Kubernetes (are additional PRs required)?
* Is there a scenario where a GET could be harmful or undesirable?

**Comment by JockDaRock** [comment]
Hmmm... that is interesting.  I don't know why it wouldn't rebase, unless some changes have happened since I posted this. Might have to rework that then.

If a function expects a body and doesn't get one, the function would error out with whatever is built into the function code to handle a no-body request.  This is identical to a POST with no body.  The way I have tested this is enabling POST and GET on a flask route.  The flask function returns the same output when being used on POST and GET.  Functions that are not processing the HTTP  headers will ignore that part of the request anyway.

I have not tried it in my k8s cluster, I would need to update the gateway and give it a try

I can't imagine when GET would be harmful, but maybe I am not being creatvie enough :wink:.  Undesirable... that might be subjective.  To be honest, though, I can imagine a lot of functions that need to run **without** the need for a body, in which case running a GET on a function would be preferable.

Additional thoughts:
There needs to be support for, at the very least, running GET and POST on functions. makes sense for the reason I stated about running functions without a body.  Seems off to do a POST on a function that is not expecting a body, for no other reason than the platforms gateway prevents me from doing any other type of HTTP request.  That and if we are really trying to get a Flask or Express developer to take it seriously, having to put an additional component, like NGINX, in or order to rewrite incoming requests seems a bit too hacky.  That and a dev who is using Flask/Express a website or API function may not have a say on what is deployed in front of it.  which would make it more likely they won't use OpenFaaS.

File: gateway/handlers/proxy.go (hunk 1 above)
Comment (JockDaRock) [reply gateway/handlers/proxy.go]: Yes, that is true, if you would prefer that I can make that change :smile:

**Comment by JockDaRock** [comment]
any other thoughts on the PR @alexellis? 😃

**Comment by alexellis** [comment]
We have a rebasing issue.. so might need to raise this again.

**Pull request closed without merging**